File: acp-sdk-go/discovery.go
package acp

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"
)

type cachedDocument struct {
	IdentityDocument map[string]any `json:"identity_document"`
	FetchedAt        string         `json:"fetched_at"`
}

func cacheValid(identityDocument map[string]any) bool {
	validUntilRaw, ok := identityDocument["valid_until"].(string)
	if !ok {
		return false
	}
	validUntil, err := time.Parse(time.RFC3339, validUntilRaw)
	if err != nil {
		return false
	}
	return validUntil.After(time.Now().UTC())
}

func extractIdentityDocument(body map[string]any) map[string]any {
	if nested, ok := body["identity_document"].(map[string]any); ok {
		return nested
	}
	if _, hasAgent := body["agent_id"]; hasAgent {
		if _, hasKeys := body["keys"]; hasKeys {
			if _, hasService := body["service"]; hasService {
				return body
			}
		}
	}
	return nil
}

type DiscoveryClient struct {
	cachePath      string
	defaultScheme  string
	relayHints     []string
	directoryHints []string
	timeoutSeconds int
	policy         HTTPSecurityPolicy
	httpClient     *http.Client

	lock     sync.Mutex
	cache    map[string]cachedDocument
	registry map[string]map[string]any
}

func NewDiscoveryClient(
	cachePath string,
	defaultScheme string,
	relayHints []string,
	directoryHints []string,
	timeoutSeconds int,
	policy HTTPSecurityPolicy,
) (*DiscoveryClient, error) {
	httpClient, err := BuildHTTPClient(policy, timeoutSeconds)
	if err != nil {
		return nil, err
	}
	client := &DiscoveryClient{
		cachePath:      strings.TrimSpace(cachePath),
		defaultScheme:  firstNonBlankString(strings.TrimSpace(defaultScheme), "https"),
		relayHints:     append([]string{}, relayHints...),
		directoryHints: append([]string{}, directoryHints...),
		timeoutSeconds: maxInt(timeoutSeconds, 1),
		policy:         policy,
		httpClient:     httpClient,
		cache:          map[string]cachedDocument{},
		registry:       map[string]map[string]any{},
	}
	client.loadCache()
	return client, nil
}

func (client *DiscoveryClient) Seed(identityDocument map[string]any) {
	agentID, ok := identityDocument["agent_id"].(string)
	if !ok || strings.TrimSpace(agentID) == "" {
		return
	}
	client.lock.Lock()
	client.cache[agentID] = cachedDocument{
		IdentityDocument: identityDocument,
		FetchedAt:        time.Now().UTC().Format(time.RFC3339),
	}
	client.lock.Unlock()
	client.persistCache()
}

func (client *DiscoveryClient) RegisterIdentityDocument(identityDocument map[string]any) error {
	agentID, ok := identityDocument["agent_id"].(string)
	if !ok || strings.TrimSpace(agentID) == "" {
		return ValidationError("Identity document missing agent_id")
	}
	client.lock.Lock()
	client.registry[agentID] = identityDocument
	client.cache[agentID] = cachedDocument{
		IdentityDocument: identityDocument,
		FetchedAt:        time.Now().UTC().Format(time.RFC3339),
	}
	client.lock.Unlock()
	client.persistCache()
	return nil
}

func (client *DiscoveryClient) Resolve(agentID string) (map[string]any, error) {
	client.lock.Lock()
	if registered, ok := client.registry[agentID]; ok {
		client.lock.Unlock()
		return registered, nil
	}
	client.lock.Unlock()
	if cached := client.tryCache(agentID); cached != nil {
		return cached, nil
	}
	if wellKnown := client.tryWellKnown(agentID); wellKnown != nil {
		client.cacheIdentity(agentID, wellKnown)
		return wellKnown, nil
	}
	if relayLookup := client.tryHintLookups(client.relayHints, agentID); relayLookup != nil {
		client.cacheIdentity(agentID, relayLookup)
		return relayLookup, nil
	}
	if directoryLookup := client.tryHintLookups(client.directoryHints, agentID); directoryLookup != nil {
		client.cacheIdentity(agentID, directoryLookup)
		return directoryLookup, nil
	}
	return nil, DiscoveryError(fmt.Sprintf("Unable to resolve identity document for %s", agentID))
}

func (client *DiscoveryClient) ResolveWellKnown(baseURL string, expectedAgentID string) (map[string]any, error) {
	wellKnownURL, err := WellKnownURLFromBase(baseURL)
	if err != nil {
		return nil, err
	}
	resolved := client.resolveWellKnownURL(wellKnownURL, expectedAgentID)
	if resolved == nil {
		return nil, DiscoveryError(fmt.Sprintf("Unable to resolve well-known metadata from %s", wellKnownURL))
	}
	identityDocument, _ := resolved["identity_document"].(map[string]any)
	agentID, _ := identityDocument["agent_id"].(string)
	if strings.TrimSpace(agentID) == "" {
		return nil, DiscoveryError("Well-known discovery returned identity document without agent_id")
	}
	client.cacheIdentity(agentID, identityDocument)
	resolved["well_known_url"] = wellKnownURL
	return resolved, nil
}

func (client *DiscoveryClient) tryCache(agentID string) map[string]any {
	client.lock.Lock()
	cached, ok := client.cache[agentID]
	client.lock.Unlock()
	if !ok {
		return nil
	}
	if cacheValid(cached.IdentityDocument) {
		return cached.IdentityDocument
	}
	client.lock.Lock()
	delete(client.cache, agentID)
	client.lock.Unlock()
	client.persistCache()
	return nil
}

func (client *DiscoveryClient) tryWellKnown(agentID string) map[string]any {
	parts, err := ParseAgentID(agentID)
	if err != nil || strings.TrimSpace(parts.Domain) == "" {
		return nil
	}
	wellKnownURL := fmt.Sprintf("%s://%s/.well-known/acp", client.defaultScheme, parts.Domain)
	resolved := client.resolveWellKnownURL(wellKnownURL, agentID)
	if resolved == nil {
		return nil
	}
	identityDocument, _ := resolved["identity_document"].(map[string]any)
	return identityDocument
}

func (client *DiscoveryClient) tryHintLookups(hints []string, agentID string) map[string]any {
	for _, hint := range hints {
		endpoint := strings.TrimRight(hint, "/") + "/discover"
		body := client.fetchJSON(endpoint, map[string]string{"agent_id": agentID}, "Discovery hint lookup")
		if body == nil {
			continue
		}
		identityDocument := extractIdentityDocument(body)
		if identityDocument != nil && VerifyIdentityDocument(identityDocument) {
			return identityDocument
		}
	}
	return nil
}

func (client *DiscoveryClient) resolveWellKnownURL(wellKnownURL string, expectedAgentID string) map[string]any {
	body := client.fetchJSON(wellKnownURL, nil, "Discovery .well-known lookup")
	if body == nil {
		return nil
	}
	wellKnown, err := ParseWellKnownDocument(body)
	if err != nil {
		return nil
	}
	if expectedAgentID != "" {
		if agentID, _ := wellKnown["agent_id"].(string); agentID != expectedAgentID {
			return nil
		}
	}
	identityReference, err := ResolveIdentityDocumentReference(wellKnown, wellKnownURL)
	if err != nil {
		return nil
	}
	identityBody := client.fetchJSON(identityReference, nil, "Discovery identity document lookup")
	if identityBody == nil {
		return nil
	}
	identityDocument := extractIdentityDocument(identityBody)
	if identityDocument == nil || !VerifyIdentityDocument(identityDocument) {
		return nil
	}
	if expectedAgentID != "" {
		if agentID, _ := identityDocument["agent_id"].(string); agentID != expectedAgentID {
			return nil
		}
	}
	return map[string]any{
		"well_known":        wellKnown,
		"identity_document": identityDocument,
	}
}

func (client *DiscoveryClient) fetchJSON(rawURL string, query map[string]string, context string) map[string]any {
	parsed, err := ValidateHTTPURL(rawURL, client.policy.AllowInsecureHTTP, client.policy.MTLSEnabled, context)
	if err != nil {
		return nil
	}
	if len(query) > 0 {
		values := parsed.Query()
		for key, value := range query {
			values.Set(key, value)
		}
		parsed.RawQuery = values.Encode()
	}
	request, err := http.NewRequest(http.MethodGet, parsed.String(), nil)
	if err != nil {
		return nil
	}
	response, err := client.httpClient.Do(request)
	if err != nil {
		return nil
	}
	defer response.Body.Close()
	if response.StatusCode != http.StatusOK {
		return nil
	}
	rawBody, err := io.ReadAll(response.Body)
	if err != nil {
		return nil
	}
	parsedBody, err := ParseJSONMap(rawBody)
	if err != nil {
		return nil
	}
	return parsedBody
}

func (client *DiscoveryClient) cacheIdentity(agentID string, identityDocument map[string]any) {
	client.lock.Lock()
	client.cache[agentID] = cachedDocument{
		IdentityDocument: identityDocument,
		FetchedAt:        time.Now().UTC().Format(time.RFC3339),
	}
	client.lock.Unlock()
	client.persistCache()
}

func (client *DiscoveryClient) loadCache() {
	if client.cachePath == "" {
		return
	}
	raw, err := os.ReadFile(client.cachePath)
	if err != nil {
		return
	}
	parsed := map[string]cachedDocument{}
	if err := json.Unmarshal(raw, &parsed); err != nil {
		return
	}
	client.lock.Lock()
	for agentID, entry := range parsed {
		if entry.IdentityDocument != nil {
			client.cache[agentID] = entry
		}
	}
	client.lock.Unlock()
}

func (client *DiscoveryClient) persistCache() {
	if client.cachePath == "" {
		return
	}
	client.lock.Lock()
	snapshot := map[string]cachedDocument{}
	for key, value := range client.cache {
		snapshot[key] = value
	}
	client.lock.Unlock()
	data, err := json.MarshalIndent(snapshot, "", "  ")
	if err != nil {
		return
	}
	_ = os.MkdirAll(filepath.Dir(client.cachePath), 0o755)
	_ = os.WriteFile(client.cachePath, data, 0o644)
}

func withQuery(rawURL string, query map[string]string) string {
	if len(query) == 0 {
		return rawURL
	}
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return rawURL
	}
	values := parsed.Query()
	for key, value := range query {
		values.Set(key, value)
	}
	parsed.RawQuery = values.Encode()
	return parsed.String()
}

File: sdks/typescript/src/discovery.ts
/*
 * Copyright 2026 ACP Project
 * Licensed under the Apache License, Version 2.0
 * See LICENSE file for details.
 */

import { existsSync, readFileSync, writeFileSync } from "node:fs";
import { dirname } from "node:path";
import { mkdirSync } from "node:fs";
import { discoveryError, validationError } from "./errors";
import { HttpSecurityPolicy, buildFetchOptions, validateHttpUrl } from "./httpSecurity";
import { parseAgentId, verifyIdentityDocument } from "./identity";
import { JsonMap, JsonValue, parseJsonMap, toJsonMap } from "./jsonSupport";
import {
  parseWellKnownDocument,
  resolveIdentityDocumentReference,
  wellKnownUrlFromBase
} from "./wellKnown";

interface CachedDocument {
  identity_document: JsonMap;
  fetched_at: string;
}

function cacheValid(identityDocument: JsonMap): boolean {
  const validUntil = identityDocument.valid_until;
  if (typeof validUntil !== "string") {
    return false;
  }
  const expires = Date.parse(validUntil);
  return Number.isFinite(expires) && expires > Date.now();
}

function extractIdentityDocument(body: JsonMap): JsonMap | undefined {
  if (body.identity_document && typeof body.identity_document === "object" && !Array.isArray(body.identity_document)) {
    return body.identity_document as JsonMap;
  }
  if (body.agent_id && body.keys && body.service) {
    return body;
  }
  return undefined;
}

export class DiscoveryClient {
  private readonly cache = new Map<string, CachedDocument>();
  private readonly registry = new Map<string, JsonMap>();
  private readonly fetchOptions: ReturnType<typeof buildFetchOptions>;

  public constructor(
    private readonly cache_path: string | undefined,
    private readonly default_scheme = "https",
    private readonly relay_hints: string[] = [],
    private readonly enterprise_directory_hints: string[] = [],
    private readonly timeout_seconds = 10,
    private readonly policy: HttpSecurityPolicy = {
      allow_insecure_http: false,
      allow_insecure_tls: false,
      mtls_enabled: false
    }
  ) {
    this.fetchOptions = buildFetchOptions(policy);
    this.loadCache();
  }

  public seed(identityDocument: JsonMap): void {
    const agentId = identityDocument.agent_id;
    if (typeof agentId !== "string" || !agentId.trim()) {
      return;
    }
    this.cache.set(agentId, {
      identity_document: identityDocument,
      fetched_at: new Date().toISOString()
    });
    this.persistCache();
  }

  public registerIdentityDocument(identityDocument: JsonMap): void {
    const agentId = identityDocument.agent_id;
    if (typeof agentId !== "string" || !agentId.trim()) {
      throw validationError("Identity document missing agent_id");
    }
    this.registry.set(agentId, identityDocument);
    this.cache.set(agentId, {
      identity_document: identityDocument,
      fetched_at: new Date().toISOString()
    });
    this.persistCache();
  }

  public async resolve(agentId: string): Promise<JsonMap> {
    if (this.registry.has(agentId)) {
      return this.registry.get(agentId) as JsonMap;
    }
    const cached = this.tryCache(agentId);
    if (cached) {
      return cached;
    }
    const wellKnown = await this.tryWellKnown(agentId);
    if (wellKnown) {
      this.cacheIdentity(agentId, wellKnown);
      return wellKnown;
    }
    const relayLookup = await this.tryHintLookups(this.relay_hints, agentId);
    if (relayLookup) {
      this.cacheIdentity(agentId, relayLookup);
      return relayLookup;
    }
    const directoryLookup = await this.tryHintLookups(this.enterprise_directory_hints, agentId);
    if (directoryLookup) {
      this.cacheIdentity(agentId, directoryLookup);
      return directoryLookup;
    }
    throw discoveryError(`Unable to resolve identity document for ${agentId}`);
  }

  public async resolveWellKnown(baseUrl: string, expectedAgentId?: string): Promise<JsonMap> {
    const wellKnownUrl = wellKnownUrlFromBase(baseUrl);
    const resolved = await this.resolveWellKnownUrl(wellKnownUrl, expectedAgentId);
    if (!resolved) {
      throw discoveryError(`Unable to resolve well-known metadata from ${wellKnownUrl}`);
    }
    const identityDocument = toJsonMap(resolved.identity_document as JsonValue);
    const agentId = identityDocument.agent_id;
    if (typeof agentId !== "string" || !agentId.trim()) {
      throw discoveryError("Well-known discovery returned identity document without agent_id");
    }
    this.cacheIdentity(agentId, identityDocument);
    return {
      ...resolved,
      well_known_url: wellKnownUrl
    };
  }

  private tryCache(agentId: string): JsonMap | undefined {
    const cached = this.cache.get(agentId);
    if (!cached) {
      return undefined;
    }
    if (cacheValid(cached.identity_document)) {
      return cached.identity_document;
    }
    this.cache.delete(agentId);
    this.persistCache();
    return undefined;
  }

  private async tryWellKnown(agentId: string): Promise<JsonMap | undefined> {
    const parts = parseAgentId(agentId);
    if (!parts.domain) {
      return undefined;
    }
    const wellKnownUrl = `${this.default_scheme}://${parts.domain}/.well-known/acp`;
    const resolved = await this.resolveWellKnownUrl(wellKnownUrl, agentId);
    if (!resolved) {
      return undefined;
    }
    return toJsonMap(resolved.identity_document as JsonValue);
  }

  private async tryHintLookups(hints: string[], agentId: string): Promise<JsonMap | undefined> {
    for (const hint of hints) {
      const url = `${hint.replace(/\/+$/, "")}/discover`;
      const body = await this.fetchJson(url, [["agent_id", agentId]], "Discovery hint lookup");
      if (!body) {
        continue;
      }
      const identityDocument = extractIdentityDocument(body);
      if (identityDocument && verifyIdentityDocument(identityDocument)) {
        return identityDocument;
      }
    }
    return undefined;
  }

  private async resolveWellKnownUrl(
    wellKnownUrl: string,
    expectedAgentId?: string
  ): Promise<JsonMap | undefined> {
    const body = await this.fetchJson(wellKnownUrl, undefined, "Discovery .well-known lookup");
    if (!body) {
      return undefined;
    }
    let wellKnown: JsonMap;
    try {
      wellKnown = parseWellKnownDocument(body);
    } catch {
      return undefined;
    }
    if (expectedAgentId && wellKnown.agent_id !== expectedAgentId) {
      return undefined;
    }
    const identityReference = resolveIdentityDocumentReference(wellKnown, wellKnownUrl);
    const identityBody = await this.fetchJson(
      identityReference,
      undefined,
      "Discovery identity document lookup"
    );
    if (!identityBody) {
      return undefined;
    }
    const identityDocument = extractIdentityDocument(identityBody);
    if (!identityDocument || !verifyIdentityDocument(identityDocument)) {
      return undefined;
    }
    if (expectedAgentId && identityDocument.agent_id !== expectedAgentId) {
      return undefined;
    }
    return {
      well_known: wellKnown,
      identity_document: identityDocument
    };
  }

  private async fetchJson(
    rawUrl: string,
    query: Array<[string, string]> | undefined,
    context: string
  ): Promise<JsonMap | undefined> {
    const parsed = validateHttpUrl(
      rawUrl,
      this.policy.allow_insecure_http,
      this.policy.mtls_enabled,
      context
    );
    if (query && query.length > 0) {
      for (const [key, value] of query) {
        parsed.searchParams.set(key, value);
      }
    }
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), Math.max(1, this.timeout_seconds) * 1000);
    try {
      const response = await fetch(parsed.toString(), {
        method: "GET",
        signal: controller.signal,
        ...this.fetchOptions
      } as RequestInit);
      if (response.status !== 200) {
        return undefined;
      }
      const rawBody = await response.text();
      return parseJsonMap(rawBody);
    } catch {
      return undefined;
    } finally {
      clearTimeout(timeout);
    }
  }

  private cacheIdentity(agentId: string, identityDocument: JsonMap): void {
    this.cache.set(agentId, {
      identity_document: identityDocument,
      fetched_at: new Date().toISOString()
    });
    this.persistCache();
  }

  private loadCache(): void {
    if (!this.cache_path || !existsSync(this.cache_path)) {
      return;
    }
    try {
      const raw = parseJsonMap(readFileSync(this.cache_path, "utf-8"));
      for (const [agentId, value] of Object.entries(raw)) {
        if (value && typeof value === "object" && !Array.isArray(value)) {
          const entry = value as JsonMap;
          const identityDocument = extractIdentityDocument(entry);
          if (identityDocument) {
            this.cache.set(agentId, {
              identity_document: identityDocument,
              fetched_at: typeof entry.fetched_at === "string" ? entry.fetched_at : new Date().toISOString()
            });
          }
        }
      }
    } catch {
      // Invalid cache is ignored.
    }
  }

  private persistCache(): void {
    if (!this.cache_path) {
      return;
    }
    try {
      mkdirSync(dirname(this.cache_path), { recursive: true });
      const serialized: JsonMap = {};
      for (const [agentId, entry] of this.cache.entries()) {
        serialized[agentId] = {
          identity_document: entry.identity_document,
          fetched_at: entry.fetched_at
        };
      }
      writeFileSync(this.cache_path, JSON.stringify(serialized, null, 2), "utf-8");
    } catch {
      // Cache persistence failures are tolerated.
    }
  }
}
